### PR TITLE
OpenTitan Earlgrey pinmux implementations

### PR DESCRIPTION
**Pull request by mazurek-michal**
### Pull Request Overview

* Update Earlgrey auto-generated files
  * Remove redundant chip name form definition naming
  * Re-licencing generated data for TockOS
* Add implementations of OpenTitan pinmux
   *  Add support for pad attribute configurations
   *  Add MIO pad/peripheral selections

### Testing Strategy

Manual test with  Hyperdebug + CW310 FPGA board and custom application.

### TODO

- [x] Missing spell check
- [x] Add  conversion operator for MIO
- [x] Integration with GPIO driver (WIP) https://github.com/tock/tock/pull/3710 
- [ ] Next PR - sleep and wake up functionality

### Help Wanted 

- [ ] Discussion -  common pinmux HIL traits

### Documentation Updated

- [x]  no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: chips/earlgrey/src/registers/top_earlgrey.rs [hunk 1]
```diff
@@ -2,10 +2,10 @@
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 // Copyright lowRISC contributors 2023.
 
-// Built for Earlgrey-M2.5.1-RC1-389-g21ce4e9761
-// https://github.com/lowRISC/opentitan/tree/21ce4e9761abdf5c919b46e5ae64a5a8e24992f7
+// Built for Earlgrey-M2.5.2-RC0-904-g9a5280ee33
+// https://github.com/lowRISC/opentitan/tree/9a5280ee3331bc20f2659eee37a56f270454bdcf
```
Comment (twilfredo) [inline chips/earlgrey/src/registers/top_earlgrey.rs:6]: We should also update the [Makefile sha](https://github.com/tock/tock/blob/3df5b00e00121f645740ec39441f93cb101feec1/boards/opentitan/earlgrey-cw310/Makefile#L19). It looks like there's a mismatch already too.

File: chips/earlgrey/src/pinmux.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+use crate::registers::pinmux_regs::{
+    PinmuxRegisters, DIO_PAD_ATTR_REGWEN, MIO_OUTSEL_REGWEN, MIO_PAD_ATTR_REGWEN,
+    MIO_PERIPH_INSEL_REGWEN,
+};
+use crate::registers::top_earlgrey::{
+    DirectPads, MuxedPads, PinmuxInsel, PinmuxOutsel, PinmuxPeripheralIn, PINMUX_AON_BASE_ADDR,
+};
+
+// To avoid code duplication for MIO/DIO we introcude
```
Comment (twilfredo) [inline chips/earlgrey/src/pinmux.rs:20]: s/introcude/introduce

File: chips/earlgrey/src/pinmux.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+// one register layout for bouth type of PAD. In feature this code
```
Comment (twilfredo) [inline chips/earlgrey/src/pinmux.rs:21]: 1. s/type of PAD/types of PADS ?
2. s/In feature/In future ?

File: chips/earlgrey/src/pinmux.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+// should be replaced by oficall improved auto generated definitions.
```
Comment (twilfredo) [inline chips/earlgrey/src/pinmux.rs:22]: s/oficall/officially

File: chips/earlgrey/src/pinmux.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+// //TODO add links to gound true
```
Comment (twilfredo) [inline chips/earlgrey/src/pinmux.rs:23]: what is "gound true"?

File: chips/earlgrey/src/pinmux.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+type PadAttribute = LocalRegisterCopy<u32, PAD_ATTR::Register>;
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Pad {
+    Mio(MuxedPads),
+    Dio(DirectPads),
+}
+
+impl Pad {
+    // Extract value of attricttes using common layout
```
Comment (twilfredo) [inline chips/earlgrey/src/pinmux.rs:50]: s/"Extract value of attricttes using common layout"/"Extract the value of attributes using a common layout"

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (bradjc) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: I think it's ok if those SHAs don't match. Since the HW isn't changing, we shouldn't need to update these definitions for correctness, but changes like this renaming might come up time-to-time.

**Review by bradjc — DISMISSED** [review]

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (mazurek-michal) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: This commit is based on `upstream/master OpenTitan` version of file  `top_earlgrey.rs`.
If it will be possible i would like to stay on top of Earlgrey ES branch because this branch will be used for HW samples.  Changes in `top_earlgrey.rs`  are pure syntax sugar.

Comment (cfrantz) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: As long as this file (top_earlgrey.rs) came from the `earlgrey_es` branch, then we don't need to update the SHA referenced in the Makefile.  There are no hardware changes to anything on the `earlgrey_es` branch (apart from the already-known FPGA clocking changes).

The SHA in the Makefile (`7df72d61...`) refers to the correct hardware version and the bitstream one should download from https://storage.googleapis.com/opentitan-bitstreams/

**Review by twilfredo — APPROVED** [review]

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (alistair23) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: Why not just update them both at the same time? If there are no HW changes it's a very simple change in Tock and that way everything is clear and in sync.

It seems strange to be using different hardware revisions for the registers and the bitstreams

File: chips/earlgrey/src/pinmux.rs [hunk 7]
```diff
@@ -0,0 +1,208 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
+//! Mux'ing between physical pads and GPIO or other peripherals.
```
Comment (alistair23) [inline chips/earlgrey/src/pinmux.rs:5]: Can we add tests for these? How do we verify that they are working

File: chips/earlgrey/src/pinmux.rs [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+type PadAttribute = LocalRegisterCopy<u32, PAD_ATTR::Register>;
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Pad {
+    Mio(MuxedPads),
+    Dio(DirectPads),
+}
+
+impl Pad {
```
Comment (alistair23) [inline chips/earlgrey/src/pinmux.rs:54]: Can we  use the `gpio::Configure` HIL here instead?

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: chips/earlgrey/src/pinmux.rs (hunk 8 above)
Comment (mazurek-michal) [reply chips/earlgrey/src/pinmux.rs:54]: You need information from mutex selection for making a working GPIO pin.
I think the best pace to aggregate all configurations is GPIO struct.

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (jrvanwhy) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: The bitstream should come from the `earlgrey_es` branch, as that branch is being used for hardware production.

The register definition generation tool is not yet stable. We are still refining it (as you can see in this PR), and plan to continue to work on it. For that reason, the registers should be generated using the `master` branch.

Which means the exact commits won't match, but that is fine.

Comment (alistair23) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: Why not just apply the change to the same place and update all of the SHAs here? What happens when a hardware change goes into `master` and then gets propagated here but then doesn't match what we are flashing to the board?

File: chips/earlgrey/src/pinmux.rs (hunk 7 above)
Comment (alistair23) [reply chips/earlgrey/src/pinmux.rs:5]: I understand that a pinmux affects the GPIO hardware. We can still add tests that change those and can then verify them externally. It's what we do with LEDs on some platforms for example.

Otherwise we have a driver that isn't called, so it's very difficult to test and catch regressions

Comment (alistair23) [reply chips/earlgrey/src/pinmux.rs:5]: How did you end up testing this?

File: chips/earlgrey/src/pinmux.rs (hunk 8 above)
Comment (alistair23) [reply chips/earlgrey/src/pinmux.rs:54]: Do you mean the current GPIO implementation will call these?

Comment (mazurek-michal) [reply chips/earlgrey/src/pinmux.rs:54]: Yes

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (mazurek-michal) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: I agree with @jrvanwhy - we should keep bitstream SHA on  `earlgrey_es`  and update software related things to upstream `master` branch if they don't collide with hardware changes.
Hardware is more that end user interface (register layout), using upstream version of bitstream can cause some trouble in debugging issue, because you basically debug different implementation of hardware.

However, due to the imperfections of the data generation process for Rust, it is best to use a upstream version. The version with `earlgrey_es` does not yet implement the Copy,Clone, PartialEq,Eq traits also we have repetitions of chip name prefix in constans, with can cause problem with newer chip version.

Comment (bradjc) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different. That actually makes sense to me, if the HW is frozen then the tools shouldn't need to be fixed to the same commit. Much in the same way that libtock-c and the kernel evolve separately.

Comment (jrvanwhy) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: > Why not just apply the change to the same place and update all of the SHAs here? What happens when a hardware change goes into `master` and then gets propagated here but then doesn't match what we are flashing to the board?

Cherrypicking software changes into `earlgrey_es` requires extra effort, and making a habit of it would prevent the `earlgrey_es` branch from stabilizing.

In the long run, we may need other solutions for generating Earlgrey registers (such as by running a latest-`master` register generation tool on register definitions from the `earlgrey_es` branch), but we're not there yet and using `master` to generate register definitions is what makes sense now.

Comment (cfrantz) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: > Why not just update them both at the same time? If there are no HW changes it's a very simple change in Tock and that way everything is clear and in sync.
>
>It seems strange to be using different hardware revisions for the registers and the bitstreams

The have been some software and infrastructure improvements on `master` that make cherry-picking the reggen/topgen updates onto `earlgrey_es` impractical.  Since HW changes are nearly forbidden[^1] on `master`, we're going to generate these constants from master.

We want to keep the SHA in the makefile the same because that SHA refers to the version of the codebase that would construct the exact bitstream, verilator model, test_rom and mask_rom that are being used for the chip tape-out.  If you're running tock on the FPGA or in a simulator, **that** is the version you want.

[^1] We still permit ECOs on `master`.  The ECOs are typically done on `earlgrey_es` in a way that is compatible with the floor-plan and physical layout of the chip.  Those ECOs are then re-done on `master` the "right" way (rather than bound by the tape-out constraints in `earlgrey_es`).

File: chips/earlgrey/src/pinmux.rs [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    /// Prohibits any further changes to input/output/open-drain or pullup configuration.
+    pub fn lock_pad_attributes(&self) {
+        match *self {
+            Self::Mio(mio) => PINMUX_BASE.mio_pad_attr_regwen[(mio as u32) as usize]
+                .write(MIO_PAD_ATTR_REGWEN::EN_0::CLEAR),
+            Self::Dio(dio) => PINMUX_BASE.dio_pad_attr_regwen[(dio as u32) as usize]
+                .write(DIO_PAD_ATTR_REGWEN::EN_0::CLEAR),
+        };
+    }
+}
```
Comment (jrvanwhy) [inline chips/earlgrey/src/pinmux.rs:117]: The methods inside `impl Pad` use a mix of `&self` and `self`. I think either is fine, but please make it consistent.

Comment (mazurek-michal) [reply chips/earlgrey/src/pinmux.rs:117]: Replaced to `&self`

**Review by jrvanwhy — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (alistair23) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: > Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.

How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic

> Since HW changes are nearly forbidden[^1] on `master`, we're going to generate these constants from master.

This doesn't really explain why they can't be in sync. If a hardware change happens on master the register definition will change, but the bitstream we are using won't? That seems prone to breakage.

Like you say, there isn't going to be a lot of churn on master, so why not just update it?

> That actually makes sense to me, if the HW is frozen then the tools shouldn't need to be fixed to the same commit. Much in the same way that libtock-c and the kernel evolve separately.

But the hardware isn't frozen. There has already been one change  since tapeout for FPGAs and @cfrantz  just said that some changes can go in.

libtock-c and the kernel are both written to be API compatible. A change in the hardware would generally be an "ABI breakage", so it doesn't seem exactly the same

> We want to keep the SHA in the makefile the same because that SHA refers to the version of the codebase that would construct the exact bitstream, verilator model, test_rom and mask_rom that are being used for the chip tape-out. If you're running tock on the FPGA or in a simulator, **that** is the version you want.

Isn't that not the case? We have already bumped the SHA we use in the Makefile since tapeout to get faster clock frequencies. Why can't we do the same to update register layouts.

We are already bumping the Makefile from tape out: https://github.com/tock/tock/pull/3703/files. So we are no longer targeting the bitstream from hardware production. It seems strange that it wasn't an issue for https://github.com/tock/tock/pull/3703 but it's an issue now. What am I missing here?

File: chips/earlgrey/src/pinmux.rs (hunk 8 above)
Comment (alistair23) [reply chips/earlgrey/src/pinmux.rs:54]: Ah ok. I agree that can be done separately. That is a useful bit of information to include in the commit message next time as it sets the context for the change

Comment (alistair23) [reply chips/earlgrey/src/pinmux.rs:54]: Ah, I do see it under TODO in the description

Generally TODO applies to things that need to be done in this PR before it is ready to be merged

File: chips/earlgrey/src/pinmux.rs (hunk 7 above)
Comment (mazurek-michal) [reply chips/earlgrey/src/pinmux.rs:5]: Short answer:  I used CW310 as DUT and Hyperdebug & opentitantool for GPIO stimulations.
But as you write we need some driver to call this code but it is't upstream on TockOS.

TL;DR:
I share your concerns about code quality but we currently don't have any decnet test framework in TockOS and OpenTitan. (that can test hardware feature like this).  We are working on some integration testing for TockOS in OpenTitan but it probably take some time. 

This PR https://github.com/tock/tock/pull/3710 will help with testing it on upstream TockOS. 
We can continue this topic in PR described above.

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (mazurek-michal) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: @alistair23 I don't know how much do you know about OpenTitan development process, i see this topic is hot and opinionated - we should focus more on possible solutions. 

From my point of view if this PR will be discarded current approach (`master`) will be needed https://github.com/lowRISC/opentitan/pull/19988

Other option is to rewrite code in such way that all traits will be implemented manually.

Comment (bradjc) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: > > Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.
> 
> How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic

If I'm understanding correctly, what is changing in new commits isn't the actual register meanings, but instead the tool. Updating the tool for the auto code generation shouldn't require new hardware testing.

Comment (jrvanwhy) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: > > Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.
> 
> How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic

These changes to the register definition are *software* changes, not *hardware* changes. The tool that processes the register description files (which are stable) has proved to generate better definitions. We want to use the better definitions here in Tock, so we want to generate the definitions using the tool from latest `master`.

> > Since HW changes are nearly forbidden[^1] on `master`, we're going to generate these constants from master.
> 
> This doesn't really explain why they can't be in sync. If a hardware change happens on master the register definition will change, but the bitstream we are using won't? That seems prone to breakage.

If hardware changes were to happen on master, sure, but (except for ECOs) hardware changes are banned on `master` too.

> Like you say, there isn't going to be a lot of churn on master, so why not just update it?

We could pull the bitstream from `master`, but our goal here is to run Tock on the real hardware, and `earlgrey_es` is the closest bitstream we have to that.

> > That actually makes sense to me, if the HW is frozen then the tools shouldn't need to be fixed to the same commit. Much in the same way that libtock-c and the kernel evolve separately.
> 
> But the hardware isn't frozen. There has already been one change since tapeout for FPGAs and @cfrantz just said that some changes can go in.

ECOs are very unlikely to result in register changes.

> libtock-c and the kernel are both written to be API compatible. A change in the hardware would generally be an "ABI breakage", so it doesn't seem exactly the same
> 
> > We want to keep the SHA in the makefile the same because that SHA refers to the version of the codebase that would construct the exact bitstream, verilator model, test_rom and mask_rom that are being used for the chip tape-out. If you're running tock on the FPGA or in a simulator, **that** is the version you want.
> 
> Isn't that not the case? We have already bumped the SHA we use in the Makefile since tapeout to get faster clock frequencies. Why can't we do the same to update register layouts.
> 
> We are already bumping the Makefile from tape out: https://github.com/tock/tock/pull/3703/files. So we are no longer targeting the bitstream from hardware production. It seems strange that it wasn't an issue for #3703 but it's an issue now. What am I missing here?

The SHA is #3703 is just a newer commit from the `earlgrey_es` branch. You're asking us to either copy a bunch of software changes into `earlgrey_es` or use a hardware bitstream from `master`, and both of those ideas have issues with no upside other than "the SHAs match".

Comment (jrvanwhy) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: > > > Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.
> > 
> > 
> > How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic
> 
> If I'm understanding correctly, what is changing in new commits isn't the actual register meanings, but instead the tool. Updating the tool for the auto code generation shouldn't require new hardware testing.

Yes, exactly.

Comment (alistair23) [reply chips/earlgrey/src/registers/top_earlgrey.rs:6]: I understand that the `master` OT branch has software changes for the register layout generation.

Like Johnathon just said, there might be hardware changes on master to fix issues. When/If that happens and we bump the register layout again the hardware used to generate the registers and the actual bitstream will be out of sync. I understand you expect it to be unlikey that there will be register changes, but we don't know for sure that won't happen.

Hardware wise `earlgrey_es` and `master` should be practically the same as you just said. There are no new features being developed on master, so we don't expect a lot of churn.

So what is wrong with just updating the bitstream SHA to point to master? You are worried it is out of sync? But then the registers generated would be out of sync.

> The SHA is #3703 is just a newer commit from the `earlgrey_es` branch. You're asking us to either copy a bunch of software changes into `earlgrey_es` or use a hardware bitstream from `master`, and both of those ideas have issues with no upside other than "the SHAs match".

I never said anything about copying. I just think that the register definitions we are using should match the hardware we are targeting. Which means they should be generated from the same RTL source. That way we know they are in sync.

We wouldn't generate Rustdocs for a release from a different source for example, I don't see why this is any different.

File: chips/earlgrey/src/pinmux.rs (hunk 7 above)
Comment (alistair23) [reply chips/earlgrey/src/pinmux.rs:5]: Tock on OpenTitan has a few integrated unit tests which we use to test the drivers. I think the majority of drivers have at least some tests now. We have used that for a few years to catch regressions when updating the hardware. That is a good place to start for adding tests and shouldn't take too much time.

It doesn't have something to automatically check the GPIO output though. I agree that is a limitation, but at least setting the registers and ensuring they match what we expect is something. And a lot better then nothing :)

File: chips/earlgrey/src/registers/top_earlgrey.rs (hunk 1 above)
Comment (mazurek-michal) [reply chips/earlgrey/src/registers/top_earlgrey.rs]: The PR with backport SW  changes to `earlgrey_es`was merged:
https://github.com/lowRISC/opentitan/pull/19988

Updated:
 - [x] makefile
 -  [x] register
 -  [x] soc constans
 - [ ] README
 
 let me know if i miss somthing

**Comment by bradjc** [comment]
I think we have to make a decision, and I think it is important that we not require the register definitions to be generated from the same commit as the hardware bitstream. The hardware for this version of OT is frozen and the particular SHA should stay the same. Users shouldn't have to change their OT checkout for the same hardware. However, the registers might change due to things like spelling mistakes, bugs, or adding new peripherals. They may also use SVDs in the future, but those type of changes don't affect the hardware.

**Comment by bradjc** [comment]
Re: testing: as we've discussed on the OT call several times, the OT team has a plan for testing tock as part of their larger testing framework, and that is where the effort should go. We can mark as a todo for more unit-test style tests, but in this case there just isn't much benefit since we have no way to observe the output.

**Review by bradjc — APPROVED** [review]

**Comment by mazurek-michal** [comment]
This PR was part of ongoing task described in OpenTitan issue: https://github.com/lowRISC/tock/issues/28
